### PR TITLE
Avoid undefined behavior of provided macs on EVP_MAC reinitialization

### LINKED ISSUE #17811: Reusing existing HMAC key not possible via EVP_MAC_init()
https://github.com/openssl/openssl/blob/2cc7c9b6981d683711e76c3483f813701b686eb9/providers/implementations/macs/hmac_prov.c#L172-L183

Since the above function does not call `hmac_setkey()` if no key is passed, it does not preserve the semantics of `HMAC_Init_ex()`, which allowed resetting the internal state but reusing the existing key. Is that on purpose?

### PR DESCRIPTION
**Pull request by t8m**
When the context is reinitialized, i.e. the same key should be used
we must properly reinitialize the underlying implementation.
    
However in POLY1305 case it does not make sense as this special MAC
should not reuse keys. We fail with this provided implementation
when reinitialization happens.
    
Fixes #17811

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Yeah, here: https://github.com/openssl/openssl/blob/801c638c50406c93d683c1ab8bd1d430cff4b6d0/apps/mac.c#L183-L186

The key is passed as a param not via this call.  I prefer the way you are heading but it means an API break :(

**Comment by t8m** [comment]
> The key is passed as a param not via this call. I prefer the way you are heading but it means an API break :(

I think it can be avoided - done in the fixup.

**Comment by t8m** [comment]
CI is green

File: test/evp_test.c [hunk 1]
```diff
@@ -1572,6 +1574,28 @@ static int mac_test_run_mac(EVP_TEST *t)
...
+        }
+        ERR_set_mark();
+        if (EVP_MAC_init(ctx, NULL, 0, ivparams)) {
+            ERR_clear_last_mark();
+            OPENSSL_free(got);
+            got = NULL;
+
+            goto retry;
+        }
+        /* If reinitialization fails, it is unsupported by the algorithm */
```
Comment (mattcaswell) [inline test/evp_test.c:1596]: Hmmm. This means the test is going to miss bugs where the initialisation fails when it should succeed. Perhaps we should have a list of algorithms that a re known to fail. If reinit fails for an algorithm that is expected to succeed then the test fails.

Comment (t8m) [reply test/evp_test.c:1596]: Yeah, this is currently the Poly1305 only. I think I can add an option for that.

Comment (t8m) [reply test/evp_test.c:1611]: Done

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for the reviews.

**Pull request closed without merging**